### ISSUE TITLE
Merge Payment Request Architecture with Capabilities Document

### ISSUE THREAD

**Issue by msporny** [w3c/browser-payment-api#42]
Migrating from https://github.com/w3c/webpayments/issues/22:

From the [Payment Request Architecture](http://wicg.github.io/paymentrequest/specs/architecture.html#abstract) document:

> The Payment Request architecture is designed to separate different concerns of the Payment Request system into different specifications so that they can be discussed and moved forward independently. This document describes the architecture and explains what different specifications may be created and how the relate to each other. [[link](http://wicg.github.io/paymentrequest/specs/architecture.html#abstract)]

It feels like the Payment Request Architecture and the Web Payments IG's Capabilities document overlap (based on the desired goals for the capabilities document). I agree that we need a document like the Payment Request Architecture, but think that it should be placed into the architecture document (capabilities) the WPIG agreed upon.

It's also confusing because it seems like there is also overlap with @adrianhopebailie's architecture document here: https://github.com/w3c/webpayments/wiki/A-Payments-Initiation-Architecture-for-the-Web

Spec refs:
http://wicg.github.io/paymentrequest/specs/architecture.html#abstract

**Comment by ianbjacobs** [comment]
-1 to merging this architecture document and the capabilities document, which has not been edited in about 6 months.

Regarding the arch wiki and the spec, please see this proposal, which includes migrating information from the wiki to the spec, then reducing the wiki:
 https://lists.w3.org/Archives/Public/public-payments-wg/2016Mar/0087.html

This does not preclude other architectural statements. The purpose of the arch spec (today) is to explain the set of documents that fit together around the paymentrequest API.

Ian

**Comment by msporny** [comment]
+1 to migrating information from the wiki to the spec.

If the arch spec is specific to the set of documents around paymentrequest API, it's not the "Payment Request Architecture" specification, it's the "Browser Payments API Design Document".

The document that @adrianba put together has a much more narrow scope and focuses on the browser, which is fine, but the title and front-matter should reflect that reality. Browsers are not the only participants in the ecosystem we're trying to build.

**Comment by ianbjacobs** [comment]
@msporny 

I changed the title to "Payment Request API Architecture" which I think accomplishes the narrowing of scope.

As I mentioned [1], my next editorial act if the arch document is approved is to shrink the wiki, leaving bits that are not addressed in the arch spec.

Ian

[1] https://lists.w3.org/Archives/Public/public-payments-wg/2016Mar/0087.html

**Comment by dlongley** [comment]
@ianbjacobs,

We will have more than one API for "payment requests", so I don't know that "Payment Request API Architecture" is sufficiently narrow unless that document will cover other APIs, not just the browser API.

**Comment by dlongley** [comment]
@ianbjacobs,

If it's scoped underneath the browser payment API, then I guess that's fine.

**Comment by dlongley** [comment]
After further reviewing the Payment Request API Architecture document, I think it needs to more clearly indicate that it is just the architecture for the API, not for an overall Web Payments ecosystem or vision. The title is at least better than how the document is referenced throughout itself, where the language "Payment Request Architecture" is used, not "Payment Request API Architecture".

Is the plan to create separate architecture documents for each API? Should we just wait to publish this until we can figure that out? If we're going to have a separate document for each API, why not just include it in the API document itself?

It seems to me that we should create a single architecture document that includes all of the high-level concepts for a Web Payments vision -- and not a separate document for each API. We could put out a primer for each (or some) of the APIs -- and perhaps this current document would fit better that in that mold.

**Comment by ianbjacobs** [comment]
@dlongley,

Yes, the references to the document internally should reflect this is for the API architecture.

> Is the plan to create separate architecture documents for each API? 

I don't think there is a plan in the sense that nothing has been proposed yet.

> Should we just wait to publish this until we can figure that out?

I prefer to publish this document because it helps to understand how this set of documents works together. It does not preclude later changes, later descriptions, etc.

> If we're going to have a separate document for each API, why not just include it in the API document itself?

Once we have proposals, we can decide whether they are most well-suited (e.g., separate architecture documents or within this one).

In short: I would like to publish this one with the others we currently have. This is expected to be a NOTE and therefore we have a lot of flexibility in how it evolves. Publishing it today should help readers understand the current set of documents, and does not constrain us in the future if we want to make changes.

Ian

**Comment by dlongley** [comment]
@ianbjacobs,

If this document is just about understanding the architecture for the Payment Request API, why not put that content into the API spec itself under an "Architecture" section? That seems like a simpler (and better) solution to me. If I want to know how other documents relate to the API spec, I'd like to read that in the API spec, not go to another document.

**Comment by ianbjacobs** [comment]
@dlongley,

My understanding is that the editors have chosen to break the content into (at least) four pieces: API, identifiers, basic card payments, registration. Three of those are available today. It makes sense to me to keep a "how all these things fit together" description as a separate document rather than inside one of them. 

This has been done for previous sets of specifications (e.g., OWL, PROV, RIF).

Ian

**Comment by dlongley** [comment]
@ianbjacobs,

Yeah, but this document seems to do things like explain Payment Apps -- which I would expect to be in a separate document as they relate to more than just the browser API. 

I can imagine a Web Payments Architecture doc similar to what @adrianhopebailie wrote that talks about the various concepts, such as Payment Apps, and the various APIs that are available in the ecosystem. For example, the document would indicate what Payment Apps are and that if you want to  register a web-based Payment App for use with the Payment Request API, use the Registration API to do that. The document could also say, if you want to write a Payment App and send a Payment Request Message to an HTTP endpoint for processing, use the HTTP API. If you want to accept payment on your website, call the Payment Request API. And so forth.

The scope and how all of the various pieces relate to one another is larger than a single API. It seems like much of the material that is in the Payment Request API Architecture document should be in such a document. Right now there is significant overlap/duplication which seems like it would require extra work to keep in sync in the future.

**Comment by ianbjacobs** [comment]
@dlongley,

These are reasonable points. 

I still feel we should publish it as a FPWD, continue to discuss, and evolve it as our published specs and picture become clearer.

Ian

**Comment by msporny** [comment]
@ianbjacobs wrote:

> This has been done for previous sets of specifications (e.g., OWL, PROV, RIF).

No, that is not how those WGs did this. In each of those cases the published document was done as a "Document Overview"... not an "Architecture" document. In addition, the Payment Request API Architecture document says things like this:

> ... the anticipated user experience will help explain the architecture:
> 1. At the end of shopping on a merchant site, the user pushes the “buy” button.
> 2. The merchant site calls the payment API with purchase amount, currency, accepted payment methods (e.g., Visa, Paypal, Bitcoin), and any custom data for those payment methods.

That statement is almost solely about the browser API and does not take any sort of HTTP API into account. Most of the document is like that, so it's about the Payment Request API and is not a "Document Overview".

We don't gain anything by publishing a document that has questionable benefit at best and could result in a fair amount of miscommunication in the average case. People could walk away from an FPWD reading of the document and think that all we care about is the browser (we care about much more than that). The Payment Request API Architecture document currently reads as if it is _the_ architecture document for Web Payments. It's clearly not.

Many specifications at W3C (rightly) integrate this sort of stuff into the front-matter of the specification. For example, HTML5 does this:

https://www.w3.org/TR/html5/introduction.html#a-quick-introduction-to-html

DOM4 does this:

https://www.w3.org/TR/domcore/#goals

SVG does this:

https://www.w3.org/TR/SVG/concepts.html

and so on and so forth.

I do not support FPWD publication of this document at this time but would support the following counter-proposal.

PROPOSAL: Integrate the majority of the Payment Request API Architecture document into the front-matter of the Payment Request API spec. The editors will do the work to integrate this after the Payment Request API FPWD in order to give the group enough time to review the changes.

**Comment by ianbjacobs** [comment]
Hi Manu,

Is the primary concern the use of the word "Architecture" in the title?

Ian

**Comment by msporny** [comment]
> Is the primary concern the use of the word "Architecture" in the title?

That is one of the concerns, but not the biggest one. The biggest concern is that almost the entirety of the content in the document should be moved into the Payment Request API document because that's where it belongs (the document is an overview of how the Payment Request API is designed).

**Comment by halindrome** [comment]
Hmm.  You said "majority" @msporny; are there parts that are NOT an overview of the Payment Request API design?  If so, where should those parts go?  Just trying to get a complete picture of your ask here.

**Comment by ianbjacobs** [comment]
I think it is useful to have a separate document for several reasons:
- It most likely addresses a different audience, one that wants a big picture rather than looking at code samples. It is more approachable to have a smaller document with a consistent tone for that audience, than a larger document where the majority is not for that audience. I don't think the current architecture document is a finished product but I think it is worth publishing as a FPWD.
- It is helpful to be able to evolve that document independently of the specs.
- I think it's also helpful to have a wrapper for what is likely to be a growing number of independent specifications (especially if we start to publish more payment method specs).

I don't have a strong view of the use of the word "Architecture v. "Overview" in the title.

Ian

**Comment by msporny** [comment]
Can we document why this issue was closed?

**Comment by ianbjacobs** [comment]
There has been no activity on the capabilities document in the IG for 1 year. Therefore, I think it is safe to close the issue based on the original proposal.

Ian

**Comment by adrianhopebailie** [comment]
This issue is not relevant for this list.
If the IG wants to pull the architecture content into the capabilities spec then that should be taken up in the IG.
